README.md
# Template Engine Factory

[![StyleCI](https://github.styleci.io/repos/21270731/shield?branch=master)](https://github.styleci.io/repos/21270731)
[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![ProcessWire 3](https://img.shields.io/badge/ProcessWire-3.x-orange.svg)](https://github.com/processwire/processwire)

A ProcessWire module integrating template engines such as _Twig_. It allows to render pages or individual templates
via template engine and encourages to separate logic from markup by implementing a simple _MVC_ pattern. 

* For a quick introduction, please read the [Getting Started](#getting-started) section of this readme.
* More information is available in the official [Documentation](DOCUMENTATION.md).

> Version `2.x` of this module differs from the `1.x` version in many ways. Modules providing template engines _must_ be
installed with Composer. Please take a look at the [update guide](DOCUMENTATION.md#updating-from-1x-to-2x), as the new
version introduces backwards compatibility breaks. The `1.x` version of the module is available on the [1.x branch](https://github.com/wanze/TemplateEngineFactory/tree/1.x).

## Requirements

* ProcessWire `3.0` or newer
* PHP `7.0` or newer
* Composer

## Installation

Execute the following command in the root directory of your ProcessWire installation:

```
composer require wanze/template-engine-factory:^2.0
```

This installs the module and its bundled template engine called _TemplateEngineProcessWire_. This template engine uses 
ProcessWire's internal `TemplateFile` class for rendering. Other template engines are added to the factory by installing
separate ProcessWire modules.

### Installing Twig and other template engines

Each template engine is a separate ProcessWire module. For example, if you want to use [Twig](https://github.com/wanze/TemplateEngineTwig),
execute the following command:

```
composer require wanze/template-engine-twig:^2.0
```

This will install the _TemplateEngineTwig_ module and all Twig dependencies.

> ℹ️ This module includes test dependencies. If you are installing it on production with `composer install`, make sure to
pass the `--no-dev` flag to omit autoloading any unnecessary test dependencies!.

## Configuration

The _TemplateEngineFactory_ offers the following configuration options:

* **`Template Engine`** The template engine used to render pages and templates. Any installed engine is listed here.
* **`Path to templates`** Relative path from the site directory where template files are stored. E.g. `templates/views/`
resolves to `/site/templates/views/`.
* **`Enable automatic page rendering`** Check to delegate the rendering of pages to the template engine.
You may enable or disable this behaviour for specific templates.
* **`API variable to interact with the template engine`** Enter a name for the API variable used to pass data from
the ProcessWire template (Controller) to the template engine.
* **`Enabled templates`** Restrict automatic page rendering to the templates selected here.
* **`Disabled templates`** Select templates of pages that should not automatically be rendered via template engine.
Do not use in combination with the _Enabled templates_ configuration,
either enable or disable templates.

> More configuration options might be available in the module providing a template engine, e.g. the
module _TemplateEngineTwig_ offers several configuration related to Twig.

## Available template engines

* **ProcessWire** A template engine using ProcessWire's *TemplateFile* class for rendering. This engine is bundled with
this module, but it is not installed automatically. Install the module _TemplateEngineProcessWire_ and select the 
engine in the _TemplateEngineFactory_ module configuration.
* **Twig** See: https://github.com/wanze/TemplateEngineTwig
* **Pug** See: https://github.com/dreerr/TemplateEnginePug
* **Smarty** See: https://github.com/blue-tomato/TemplateEngineSmarty
* **Mustache** See: https://github.com/blue-tomato/TemplateEngineMustache
* **Latte** See: https://github.com/daun/TemplateEngineLatte

## Getting Started

> This section assumes that Twig is used as active template engine, but the usage is exactly the same for any other
chosen template engine.

### Using the template engine to render templates

Assume the following Twig template exists in `/site/templates/views/foo.html.twig`

```twig
<h1>{{ title }}</h1>
{% if body %}
    <p>{{ body }}</p>
{% endif %}
```

The template can be rendered anywhere with the _Template Engine Factory_ module:

```php
$factory = wire('modules')->get('TemplateEngineFactory');

// Render foo.html.twig with some data.
$factory->render('foo', ['title' => 'Foo', 'body' => 'Hello World']);
```

### Automatic Page Rendering

If enabled, this feature uses the template engine to render ProcessWire pages when calling `Page::render`.
By default, the module tries to find a Twig template matching the same name as the ProcessWire template:

* `/site/templates/views/home.html.twig` corresponds to `/site/templates/home.php`
* `/site/templates/views/about.html.twig` corresponds to `/site/templates/about.php`

ProcessWire templates have access to a `$view` API variable which can be used to pass data to the template engine.
As the template engine is now responsible to output markup, ProcessWire templates can be seen as _Controllers_.
They process the request and pass data to the _View_ layer via the `$view` API variable.

**Examples**

Consider the following ProcessWire template in `/site/templates/home.php`

```php
// Form has been submitted, do some processing, send mail, save data... 
if ($input->post->form) {
  // ...
  $session->redirect('./');
}

// Forward some data to twig
$view->set('nav_items', $pages->get('/')->children());
```

The corresponding Twig template in `/site/templates/views/home.html.twig` might look like this:

```twig
<h1>{{ page.title }}</h1>

<ul class="nav">
{% for item in nav_items %}
    <li><a href="{{ item.url }}">{{ item.title }}</a></li>
{% endfor %}
</ul>

<form name="form">
    <input type="text" name="email">
    <input type="submit" value="Submit">
</form>
```

Note that the ProcessWire template does not echo out any markup. It just contains business logic and uses the `$view` API
variable to pass data to the Twig template. That's it! The most simple _MVC_ pattern available in ProcessWire. 😎


CHANGELOG.md
# Changelog

## [2.1.0] - 2021-07-04

### Added

* Add support for installation via Composer 2 ([#9](https://github.com/wanze/TemplateEngineFactory/pull/19))

## [2.0.1] - 2019-09-28

### Fixed

* Allow to disable system templates from being rendered by the active template engine, see [TemplateEngineTwig issue #21](https://github.com/wanze/TemplateEngineTwig/issues/21).

## [2.0.0] - 2019-02-13

* First release of the new major `2.x` version 🐣
* Check out the [documentation](DOCUMENTATION.md), especially the [update guide](DOCUMENTATION.md#updating-from-1x-to-2x)
if you like  to update from `1.x`.

### Added

* Add possibility to enable or disable automatic page rendering.
* Add a configuration to enable or disable automatic page rendering for templates.
* Add hooks allowing you to customize the module's behaviour. See [docs](DOCUMENTATION.md#hooks).
* Add unit tests 🎉

### Changed

* A lot of things! They are summarized in the [update guide](DOCUMENTATION.md#updating-from-1x-to-2x).

## [1.1.3] - 2018-03-18

### Fixed
* Fix `Notice: Undefined index: TemplateEngineNull` if no template is available for the current page
and debug mode is enabled.

## [1.1.2] - 2018-03-18

### Fixed
* Add a hook before `ProcessPageView::pageNotFound` to handle `Wire404Exception` thrown by controllers.
The current active template engine now receives the configured page displaying a 404 rather.
than the page throwing the 404 exception.

## [1.1.1] - 2018-03-18

### Added
* Template engines can now pass multiple key/value data via new method `TemplateEngine::setArray()`.

### Fixed
* Lower hook priority for the hook executed after `Page::render`. This makes sure that the current active
template engine returns the markup *before* other modules hooking after `Page::render`.

[2.1.0]: https://github.com/wanze/TemplateEngineFactory/releases/tag/v2.1.0
[2.0.1]: https://github.com/wanze/TemplateEngineFactory/releases/tag/v2.0.1
[2.0.0]: https://github.com/wanze/TemplateEngineFactory/releases/tag/v2.0.0
[1.1.3]: https://github.com/wanze/TemplateEngineFactory/releases/tag/v1.1.3
[1.1.2]: https://github.com/wanze/TemplateEngineFactory/releases/tag/v1.1.2
[1.1.1]: https://github.com/wanze/TemplateEngineFactory/releases/tag/v1.1.1


TemplateEngineFactory.module.php
<?php

namespace ProcessWire;

use TemplateEngineFactory\TemplateVariables;
use TemplateEngineFactory\TemplateEngineInterface;
use TemplateEngineFactory\TemplateEngineNull;
use TemplateEngineFactory\Controller;

/**
 * Provides ProcessWire integration for various template engines such as Twig.
 *
 * @see https://github.com/wanze/TemplateEngineFactory
 */
class TemplateEngineFactory extends WireData implements Module, ConfigurableModule
{
    /**
     * @var array
     */
    private static $defaultConfig = [
        'engine' => '',
        'auto_page_render' => 1,
        'api_var' => 'view',
        'enabled_templates' => [],
        'disabled_templates' => [],
        'templates_path' => 'templates/views/'
    ];

    /**
     * @var \TemplateEngineFactory\TemplateEngineInterface[]
     */
    private $engines = [];

    /**
     * @var \ProcessWire\WireArray
     */
    private $templateVariables;

    public function __construct()
    {
        parent::__construct();

        $this->templateVariables = $this->wire(new WireArray());
        $this->wire('classLoader')->addNamespace('TemplateEngineFactory', __DIR__ . '/src');
        $this->setDefaultConfig();
    }

    /**
     * @return array
     */
    public static function getModuleInfo()
    {
        return [
            'title' => 'Template Engine Factory',
            'version' => 210,
            'author' => 'Stefan Wanzenried',
            'summary' => 'Provides ProcessWire integration for various template engines such as Twig.',
            'href' => 'https://processwire.com/talk/topic/6833-module-templateenginefactory/',
            'singular' => true,
            'autoload' => true,
            'requires' => [
                'PHP>=7.0',
                'ProcessWire>=3.0',
            ],
        ];
    }

    /**
     * Initialize module by hooking into Page::render.
     */
    public function init()
    {
        if (!$this->get('auto_page_render')) {
            return;
        }

        $this->addHookBefore('Page::render', $this, 'hookBeforePageRender');
        $this->addHookAfter('Page::render', $this, 'hookAfterPageRender', ['priority' => '100.01']);
    }

    /**
     * Register a template engine.
     *
     * @param string $name
     *   The name of the template engine, e.g. "Twig" or "Plug".
     * @param \TemplateEngineFactory\TemplateEngineInterface $engine
     */
    public function registerEngine($name, TemplateEngineInterface $engine)
    {
        $this->engines[$name] = $engine;
    }

    /**
     * Get the current active template engine.
     *
     * @return \TemplateEngineFactory\TemplateEngineInterface
     */
    public function getEngine()
    {
        $name = $this->get('engine');

        return $this->engines[$name] ?? new TemplateEngineNull();
    }

    /**
     * Render the given template and data via template engine.
     *
     * @param string $template
     *   A relative path to the template file.
     * @param array $data
     *   Data passed to the template file.
     *
     * @return string
     */
    public function render($template, array $data = [])
    {
        return $this->getEngine()->render($template, $data);
    }

    /**
     * A controller wraps a ProcessWire template executing some logic and a template file of the active engine.
     *
     * @param string $controllerFile
     * @param string $templateFile
     *
     * @return \TemplateEngineFactory\Controller
     */
    public function controller($controllerFile, $templateFile)
    {
        return $this->wire(new Controller($this, $controllerFile, $templateFile));
    }

    /**
     * @return \TemplateEngineFactory\TemplateEngineInterface[]
     */
    public function getEngines()
    {
        return $this->engines;
    }

    /**
     * Hook before rendering a page.
     *
     * @param \ProcessWire\HookEvent $event
     */
    public function hookBeforePageRender(HookEvent $event)
    {
        /** @var \ProcessWire\Page $page */
        $page = $event->object;

        if (!$this->shouldRenderPage($page)) {
            return;
        }

        $this->prepareTemplateVariables($page);
    }

    /**
     * Hook after rendering a page.
     *
     * Replaces the return value of Page::render() with the output from
     * the template engine.
     *
     * @param \ProcessWire\HookEvent $event
     */
    public function hookAfterPageRender(HookEvent $event)
    {
        /** @var \ProcessWire\Page $page */
        $page = $event->object;

        if (!$this->shouldRenderPage($page)) {
            return;
        }

        $template = $this->resolveTemplate($page);
        $variables = $this->resolveTemplateVariables($page, $this->collectTemplateVariables());

        $event->return = $this->getEngine()->render($template, $variables);
    }

    /**
     * Get the name of the template that should be used to render the given page.
     *
     * By default, the module will load a template with the same name as the
     * page's (ProcessWire) template name.
     *
     * @param \ProcessWire\Page $page
     *
     * @return string
     */
    protected function ___resolveTemplate(Page $page)
    {
        return $page->get('template')->name;
    }

    /**
     * Check if the given page should be rendered via template engine.
     *
     * @param \ProcessWire\Page $page
     *
     * @return bool
     */
    protected function ___shouldRenderPage(Page $page)
    {
        // If the page is not viewable, there is nothing to render for the template engine.
        if (!$page->viewable()) {
            return false;
        }

        return $this->shouldRenderTemplate($page->get('template'));
    }

    /**
     * Get variables available in the template when rendering the given page.
     *
     * The returned array will be passed to the engine when rendering the page.
     * By default, the module includes the page being rendered as "_page".
     * Note that this page might be different from the global "page" object.
     *
     * @param \ProcessWire\Page $page
     * @param \TemplateEngineFactory\TemplateVariables $variables
     *
     * @return array
     */
    protected function ___resolveTemplateVariables(Page $page, TemplateVariables $variables)
    {
        return $variables->getArray();
    }

    /**
     * @param \ProcessWire\Template $template
     *
     * @return bool
     */
    private function shouldRenderTemplate(Template $template)
    {
        // Do not render admin pages.
        if ($template->name === 'admin') {
            return false;
        }

        // Do not render pages not having a ProcessWire template file, there is no "controller" available.
        if (!$template->filenameExists()) {
            return false;
        }

        // Check if the template is enabled or disabled.
        if (count($this->get('enabled_templates'))) {
            return in_array($template->id, $this->get('enabled_templates'));
        }

        if (count($this->get('disabled_templates'))) {
            return !in_array($template->id, $this->get('disabled_templates'));
        }

        return true;
    }

    /**
     * Initialize new template variables when rendering the given page.
     *
     * @param \ProcessWire\Page $page
     *
     * @throws \ProcessWire\WireException
     */
    private function prepareTemplateVariables(Page $page)
    {
        // Push any existing variables on the internal stack.
        $variables = $this->wire($this->get('api_var'));

        if ($variables instanceof TemplateVariables) {
            $this->templateVariables->append($variables);
        }

        $variables = $this->wire(new TemplateVariables(['_page' => $page]));

        $this->wire($this->get('api_var'), $variables);
    }

    /**
     * Get all collected template variables after rendering a page.
     *
     * The ProcessWire template aka "controller" has populated these
     * variables during the Page::render() call.
     *
     * @throws \ProcessWire\WireException
     *
     * @return \TemplateEngineFactory\TemplateVariables
     */
    private function collectTemplateVariables()
    {
        $variables = $this->wire($this->get('api_var'));

        // Restore previous variables from stack, in case of recursive rendering.
        if ($this->templateVariables->count()) {
            $this->wire($this->get('api_var'), $this->templateVariables->pop());
        }

        if ($variables instanceof TemplateVariables) {
            return $variables;
        }

        return $this->wire(new TemplateVariables());
    }

    private function setDefaultConfig()
    {
        foreach (self::$defaultConfig as $key => $value) {
            $this->set($key, $value);
        }
    }

    /**
     * @param array $data
     *
     * @throws \ProcessWire\WireException
     * @throws \ProcessWire\WirePermissionException
     *
     * @return \ProcessWire\InputfieldWrapper
     */
    public static function getModuleConfigInputfields(array $data)
    {
        $wrapper = new InputfieldWrapper();
        $data = array_merge(self::$defaultConfig, $data);

        $templates = [];
        foreach (wire('templates') as $template) {
            $label = $template->name;
            if ($template->flags & Template::flagSystem) {
                $label .= ' (system)';
            }
            $templates[$template->id] = $label;
        }

        /** @var \ProcessWire\InputfieldSelect $field */
        $field = wire('modules')->get('InputfieldSelect');
        $field->label = __('Template Engine');
        $field->description = __('Select the template engine which is used to render the pages.');
        $field->notes = __('More config options might be available in the module providing this engine.');
        $field->value = $data['engine'];
        $field->name = 'engine';
        $engines = wire('modules')->get('TemplateEngineFactory')->getEngines();
        $options = [];
        foreach (array_keys($engines) as $name) {
            $options[$name] = ucfirst($name);
        }
        $field->addOptions($options);
        $wrapper->append($field);

        /** @var \ProcessWire\InputfieldText $field */
        $field = wire('modules')->get('InputfieldText');
        $field->name = 'templates_path';
        $field->label = __('Path to templates');
        $field->description = __('Relative path from the site directory where template files are stored. E.g. `templates/views/` resolves to `/site/templates/views/`.');
        $field->value = $data['templates_path'];
        $field->required = 1;
        $wrapper->append($field);

        /** @var \ProcessWire\InputfieldCheckbox $field */
        $field = wire('modules')->get('InputfieldCheckbox');
        $field->label = __('Enable automatic page rendering');
        $field->description = __('Check to delegate the rendering of pages to the template engine. You may enable or disable this behaviour for specific templates.');
        $field->name = 'auto_page_render';
        $field->attr('checked', (bool) $data['auto_page_render']);
        $wrapper->append($field);

        $field = wire('modules')->get('InputfieldText');
        $field->label = __('API variable to interact with the template engine');
        $field->description = __('Enter a name for the API variable used to pass data from the ProcessWire template (Controller) to the template engine.');
        $field->name = 'api_var';
        $field->value = $data['api_var'];
        $field->required = 1;
        $field->showIf = 'auto_page_render=1';
        $wrapper->append($field);

        $field = wire('modules')->get('InputfieldAsmSelect');
        $field->label = __('Enabled templates');
        $field->description = __('Restrict automatic page rendering to the templates selected here.');
        $field->attr('name', 'enabled_templates');
        $field->attr('value', $data['enabled_templates']);
        $field->collapsed = Inputfield::collapsedBlank;
        $field->showIf = 'auto_page_render=1';
        $field->addOptions($templates);
        $wrapper->append($field);

        $field = wire('modules')->get('InputfieldAsmSelect');
        $field->label = __('Disabled templates');
        $field->description = __('Select templates that should **not** automatically be rendered via template engine.');
        $field->notes = __('Do not use in combination with the *Enabled templates* configuration above, either enable or disable templates.');
        $field->attr('name', 'disabled_templates');
        $field->attr('value', $data['disabled_templates']);
        $field->collapsed = Inputfield::collapsedBlank;
        $field->showIf = 'auto_page_render=1';
        $field->addOptions($templates);
        $wrapper->append($field);

        return $wrapper;
    }
}
